File: stdlib/js/hull/middleware/tests/test_csp.js
// test_csp.js. Tests for hull:middleware:csp
//
// Lua parity: same coverage as stdlib/lua/hull/middleware/tests/test_csp.lua.

import { csp } from "hull:middleware:csp";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

function assertTrue(v, msg) {
    if (!v) throw new Error(msg || "expected truthy value");
}

function assertContains(haystack, needle, msg) {
    if (!haystack || haystack.indexOf(needle) < 0)
        throw new Error((msg || "") + " expected '" + needle +
                        "' in: " + String(haystack));
}

function assertNotContains(haystack, needle, msg) {
    if (haystack && haystack.indexOf(needle) >= 0)
        throw new Error((msg || "") + " expected '" + needle +
                        "' NOT in: " + String(haystack));
}

function mockReq() { return { ctx: {} }; }

function mockRes() {
    const headers = {};
    let statusCode;
    let body;
    return {
        headersSet: headers,
        header(name, value) { headers[name] = value; },
        status(code) { statusCode = code; },
        json(data) { body = data; },
        getStatus() { return statusCode; },
        getBody() { return body; },
    };
}

// ── nonce() helper ───────────────────────────────────────────────────

test("nonce() returns a non-empty base64url string", () => {
    const n = csp.nonce();
    assertTrue(n !== null && n !== undefined);
    assertTrue(n.length >= 20, "128-bit base64url should be 22 chars");
    assertTrue(/^[A-Za-z0-9_-]+$/.test(n), "must match base64url alphabet");
});

test("two nonces differ (RNG sanity)", () => {
    assertTrue(csp.nonce() !== csp.nonce(), "successive nonces must differ");
});

// ── csp.htmx() ───────────────────────────────────────────────────────

test("htmx() sets CSP header and exposes cspNonce in ctx", () => {
    const mw = csp.htmx();
    const req = mockReq(), res = mockRes();
    const rc = mw(req, res);
    assertEq(rc, 0, "middleware should pass through");
    assertTrue(req.ctx.cspNonce !== undefined, "ctx.cspNonce must be set");
    const hdr = res.headersSet["Content-Security-Policy"];
    assertTrue(hdr !== undefined, "CSP header must be set");
    assertContains(hdr, "default-src 'self'");
    assertContains(hdr, "script-src 'self' 'nonce-");
    assertContains(hdr, "style-src 'self' 'nonce-");
    assertContains(hdr, "style-src-attr 'unsafe-inline'");
    assertContains(hdr, "frame-ancestors 'none'");
    assertContains(hdr, "base-uri 'self'");
    assertContains(hdr, "'nonce-" + req.ctx.cspNonce + "'",
                   "header nonce must equal ctx.cspNonce");
});

test("htmx() generates a fresh nonce per request", () => {
    const mw = csp.htmx();
    const req1 = mockReq(), res1 = mockRes();
    const req2 = mockReq(), res2 = mockRes();
    mw(req1, res1);
    mw(req2, res2);
    assertTrue(req1.ctx.cspNonce !== req2.ctx.cspNonce,
               "per-request nonces must differ");
});

// ── csp.strict() ─────────────────────────────────────────────────────

test("strict() omits style-src-attr 'unsafe-inline'", () => {
    const mw = csp.strict();
    const req = mockReq(), res = mockRes();
    mw(req, res);
    const hdr = res.headersSet["Content-Security-Policy"];
    assertTrue(hdr !== undefined);
    assertContains(hdr, "style-src 'self' 'nonce-");
    assertNotContains(hdr, "style-src-attr",
                      "strict profile must not include style-src-attr");
});

// ── reportOnly ───────────────────────────────────────────────────────

test("reportOnly: true uses CSP-Report-Only header", () => {
    const mw = csp.htmx({ reportOnly: true });
    const req = mockReq(), res = mockRes();
    mw(req, res);
    assertTrue(res.headersSet["Content-Security-Policy-Report-Only"] !== undefined,
               "Report-Only header must be set");
    assertEq(res.headersSet["Content-Security-Policy"], undefined,
             "regular CSP header must NOT be set in report-only mode");
});

print(`hull:middleware:csp: ${pass} passed, ${fail} failed`);
if (fail > 0) throw new Error("csp tests failed");

File: stdlib/lua/hull/web/middleware/tests/test_csp.lua
-- test_csp.lua. Tests for hull.web.middleware.csp

local csp = require('hull.web.middleware.csp')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

local function assert_true(v, msg)
    if not v then error(msg or "expected truthy value") end
end

local function assert_contains(haystack, needle, msg)
    if not haystack or not string.find(haystack, needle, 1, true) then
        error((msg or "") .. " expected '" .. needle ..
              "' in: " .. tostring(haystack))
    end
end

local function assert_not_contains(haystack, needle, msg)
    if haystack and string.find(haystack, needle, 1, true) then
        error((msg or "") .. " expected '" .. needle ..
              "' NOT in: " .. tostring(haystack))
    end
end

-- Mock request + response. Mirror the parts the middleware uses.
local function mock_req()
    return { ctx = {} }
end

local function mock_res()
    local headers = {}
    local status_code
    local body
    return {
        headers_set = headers,
        header = function(self, name, value) headers[name] = value end,
        status = function(self, code) status_code = code end,
        json = function(self, data) body = data end,
        get_status = function() return status_code end,
        get_body = function() return body end,
    }
end

-- ── nonce() helper ───────────────────────────────────────────────────

test("nonce() returns a non-empty base64url string", function()
    local n = csp.nonce()
    assert_true(n ~= nil, "nonce should not be nil")
    assert_true(#n >= 20, "128-bit base64url should be 22 chars")
    -- base64url chars only
    assert_eq(string.match(n, "^[A-Za-z0-9_-]+$") ~= nil, true,
              "nonce must match base64url alphabet")
end)

test("two nonces are different (RNG sanity)", function()
    local n1 = csp.nonce()
    local n2 = csp.nonce()
    assert_true(n1 ~= n2, "successive nonces must differ")
end)

-- ── csp.htmx() - Pico-compatible profile ─────────────────────────────

test("htmx() sets CSP header and exposes nonce in ctx", function()
    local mw = csp.htmx()
    local req, res = mock_req(), mock_res()
    local rc = mw(req, res)
    assert_eq(rc, 0, "middleware should pass through (return 0)")
    assert_true(req.ctx.csp_nonce ~= nil, "ctx.csp_nonce must be set")
    local hdr = res.headers_set["Content-Security-Policy"]
    assert_true(hdr ~= nil, "CSP header must be set")
    assert_contains(hdr, "default-src 'self'")
    assert_contains(hdr, "script-src 'self' 'nonce-")
    assert_contains(hdr, "style-src 'self' 'nonce-")
    -- the Pico concession:
    assert_contains(hdr, "style-src-attr 'unsafe-inline'")
    assert_contains(hdr, "frame-ancestors 'none'")
    assert_contains(hdr, "base-uri 'self'")
    -- nonce in header must match the one stored in ctx
    assert_contains(hdr, "'nonce-" .. req.ctx.csp_nonce .. "'",
                    "header nonce must equal ctx.csp_nonce")
end)

test("htmx() generates a fresh nonce per request", function()
    local mw = csp.htmx()
    local req1, res1 = mock_req(), mock_res()
    local req2, res2 = mock_req(), mock_res()
    mw(req1, res1)
    mw(req2, res2)
    assert_true(req1.ctx.csp_nonce ~= req2.ctx.csp_nonce,
                "per-request nonces must differ")
end)

-- ── csp.strict() - no inline-style escape ────────────────────────────

test("strict() omits style-src-attr 'unsafe-inline'", function()
    local mw = csp.strict()
    local req, res = mock_req(), mock_res()
    mw(req, res)
    local hdr = res.headers_set["Content-Security-Policy"]
    assert_true(hdr ~= nil)
    assert_contains(hdr, "style-src 'self' 'nonce-")
    assert_not_contains(hdr, "style-src-attr",
                        "strict profile must not include style-src-attr")
end)

-- ── report_only ──────────────────────────────────────────────────────

test("report_only=true uses CSP-Report-Only header", function()
    local mw = csp.htmx({ report_only = true })
    local req, res = mock_req(), mock_res()
    mw(req, res)
    assert_true(res.headers_set["Content-Security-Policy-Report-Only"] ~= nil,
                "Report-Only header must be set")
    assert_eq(res.headers_set["Content-Security-Policy"], nil,
              "regular CSP header must NOT be set in report-only mode")
end)

print(string.format("hull.web.middleware.csp: %d passed, %d failed", pass, fail))
if fail > 0 then error("csp tests failed") end
